compiler/optimizations/inferConstRefs.cpp

```diff
@@ -0,0 +1,377 @@
+    case PRIM_QUERY:
+    case PRIM_QUERY_PARAM_FIELD:
+    case PRIM_QUERY_TYPE_FIELD:
+    case PRIM_OPTIMIZE_ARRAY_BLK_MULT:
+    case PRIM_ACTUALS_LIST:
+    case PRIM_YIELD:
+
+    case PRIM_USED_MODULES_LIST:
+
+    case PRIM_WHEN:
+    case PRIM_CAPTURE_FN_FOR_C:
+    case PRIM_CAPTURE_FN_FOR_CHPL:
+    case PRIM_CREATE_FN_TYPE:
+
+    case PRIM_NUM_FIELDS:
+    case PRIM_IS_POD:
+    case PRIM_FIELD_NUM_TO_NAME:
+    case PRIM_FIELD_NAME_TO_NUM:
+    case PRIM_FIELD_BY_NUM:
+
+    case PRIM_FORALL_LOOP:
+    case PRIM_TO_STANDALONE:
+    case PRIM_IS_REF_ITER_TYPE:
+    case PRIM_COERCE:
+    case PRIM_CALL_RESOLVES:
+    case PRIM_METHOD_CALL_RESOLVES:
+    case PRIM_GET_COMPILER_VAR:
+    case PRIM_ZIP:
+    case PRIM_REQUIRE:
+    case NUM_KNOWN_PRIMS:
+    case PRIM_ITERATOR_RECORD_FIELD_VALUE_BY_FORMAL:
+    case PRIM_BLOCK_WHILEDO_LOOP:
+    case PRIM_BLOCK_DOWHILE_LOOP:
+    case PRIM_BLOCK_FOR_LOOP:
+    case PRIM_BLOCK_C_FOR_LOOP:
+    case PRIM_ARRAY_ALLOC:
+    case PRIM_ARRAY_FREE:
+    case PRIM_ARRAY_FREE_ELTS:
+    case PRIM_STRING_COPY:
+    case PRIM_GET_END_COUNT:
+    case PRIM_SET_END_COUNT:
+    case PRIM_TO_LEADER:
+    case PRIM_TO_FOLLOWER:
+    case PRIM_DELETE:
+    case PRIM_CALL_DESTRUCTOR:
+    case PRIM_HEAP_REGISTER_GLOBAL_VAR:
+    case PRIM_HEAP_BROADCAST_GLOBAL_VARS:
+    case PRIM_PRIVATE_BROADCAST:
+    case PRIM_RT_ERROR:
+    case PRIM_RT_WARNING:
+    case PRIM_FTABLE_CALL:
+    case PRIM_VIRTUAL_METHOD_CALL:
+    case PRIM_INT_ERROR:
+      return false;
+  }
+  return false;
+}
+
+//
+// Returns 'true' if 'sym' is (or should be) a const-ref.
+// If 'sym' can be a const-ref, but is not, this function will change either
+// the intent or qual of the Symbol to const-ref.
+//
+static bool inferConstRef(Symbol* sym) {
+  INT_ASSERT(sym->isRef());
+  bool wasConstRef = sym->qualType().getQual() == QUAL_CONST_REF;
+
+  if (sym->defPoint->parentSymbol->hasFlag(FLAG_EXTERN)) {
+    return wasConstRef;
+  }
+
+  RefInfo* info = infoMap[sym];
+
+  // 'info' may be null if the argument is never used.
+  if (info == NULL || info->finalized) {
+    return wasConstRef;
+  }
+
+  bool retval = true;
+
+  while (!info->todo.empty() && retval) {
+    SymExpr* use = info->todo.front();
+
+    // By popping, recursive calls won't try to deal with the same SymExpr.
+    info->todo.pop_front();
+
+    CallExpr* call = toCallExpr(use->parentExpr);
+    INT_ASSERT(call);
+
+    CallExpr* parent = toCallExpr(call->parentExpr);
+
+    if (call->isResolved()) {
+      ArgSymbol* form = actual_to_formal(use);
+      if (!inferConstRef(form)) {
+        retval = false;
+      }
+    }
+    else if (parent && isMoveOrAssign(parent)) {
+      if (!handleRHS(parent, call, use)) {
+        retval = false;
+      }
+    }
+    else if (call->isPrimitive(PRIM_MOVE)) {
+      if (use == call->get(1) && !call->get(2)->isRef()) {
+        retval = false;
+      } else {
+        // 'use' is the RHS of a MOVE
```

Comment:
I'm further concerned about this !call->get(2)->isRef() because we could have

 PRIM_MOVE  A B

where both A and B are new-style references, and it would mean that
*A = *B.